### ISSUE TITLE
Need an official way of writing grammars for languages too complex for TextMate

### ISSUE THREAD

**Issue by tomq42** [Microsoft/vscode#216]
TextMate isn't sufficient for many languages.

We have been integrating in to the lower level, in the src/vs/languages directory and using Modes.IState and supports.TokenisationSupport. There needs to be a way of writing an extension that can do this, which at least currently there doesn't seem to be,

Thanks.

**Comment by Tyriar** [comment]
@aeschli I'm interested in the thinking behind moving away from our own tokenization in favor of tmLanguages.

**Comment by aeschli** [comment]
For performance reasons we want the tokenizers to run in the render process. As we don't want user code to run in the render process we went for declarative tokenizers.
First we had Monarch support as well, but before the API deadline decided to go fully for text mate to keep the API simple.
Also, in order for our theming support to work well, we want all tokenizers to emit TextMate tokens. 

We are aware of the limitations and problems of TextMate, and we are open to allow other types of tokenizers, but no work is planned in this area at the moment.

**Comment by tomq42** [comment]
That's a shame. It's also slightly "unfair". In the sense that it means that Microsoft can write language modes that can do things that other people can't....

There would be zero chance of a pull request to the core of vscode being accepted for a new language mode, so we are left being unable to write a sensible language mode for vscode. We have no problem in Eclipse, there's no problem in writing similar things there, they have no issue with things running in the render thread.

At least in Monarch there was a "pop state" facility, which as far as I know has no equivalent in TextMate. In Monarch you could shift states to an explicit state, and then "pop". So you could write "subroutines". The facility made it least *possible* to write our language mode in Monarch, even if it was *much* harder work than doing it the low level way, which is what we ended up doing.

**Comment by tambry** [comment]
Tokenizing languages where a single token might be split onto multiple lines is near impossible (without very complicated workarounds) using TextMate. (see https://github.com/Microsoft/vscode-textmate/issues/32)  
Monarch would be a huge improvement, allowing better language support for more complicated/nuanced languages. Though being able to write a tokenizer using an API would be even better.

**Comment by EvgeniyPeshkov** [comment]
Hello everyone. I've developed and published syntax highlighting extension based on Tree-Sitter. It provides universal syntax coloring engine for almost any programming language (currently, C and C++ are supported OOTB). Constructing entire syntax tree, Tree-sitter efficiently overcomes all limitations of built-in TextMate grammars. It's very easy to add support for a new language. I'm planning to write HowTo in the next couple of days, but you can figure it out from source code, that is very simple and straightforward. Contributions are welcome. I've been using it by myself for a month, so I suppose it's ready for public use. At least extension can be useful until VSCode core provides stronger syntax parser.

[You can install it from VSCode Marketplace.](https://marketplace.visualstudio.com/items?itemName=evgeniypeshkov.syntax-highlighter)
[Or download .vsix package from GitHub page and install it manually.](https://github.com/EvgeniyPeshkov/syntax-highlighter/releases)
[**Please note**, that extension published in VS Code Marketplace will only work in Windows-x64](https://github.com/EvgeniyPeshkov/syntax-highlighter#distribution).
[For other operating systems, please download pre-compiled .vsix package.](https://github.com/EvgeniyPeshkov/syntax-highlighter/releases)
This will be fixed in the near future with one of the next updates.
Alternatively, you can [build extension from sources](https://github.com/EvgeniyPeshkov/syntax-highlighter/blob/master/BUILD.md).

**Comment by texastoland** [comment]
> We are aware of the limitations and problems of TextMate, and we are open to allow other types of tokenizers, but no work is planned in this area at the moment.

CC @alexdima @mjbvz

@aeschli This was 5.5 years ago 👆🏼 I understand TextMate is probably as much as a thorn in your side as for extension authors judging by some of the @mjbvz's logged issues. I'm happy to document its pain points but I imagine you already have a query in your GitHub Issues Notebooks somewhere.

Here's my understanding of the current state of things:

- The semantics token provider API lays the groundwork for mapping low-level tokens to high-level scopes for theming.
- Any additional tokenizer is waiting on @alexdima's #77140.
- Tree-sitter (used in Atom and proposed in #50140) could provide both a declarative and higher-level imperative API to the lower-level semantics token provider API. I assume that's how the [Syntax Highlighter plugin](https://github.com/microsoft/vscode/issues/216#issuecomment-494813749) works since EvgeniyPeshkov/syntax-highlighter@cadcfb9. 
- Sublime syntax (proposed in #5408) could resolve some of TextMate's limitations but with a similar support burden. Microsoft would be beholden to emulate rather than evolve it.
- Monaco's own [Monarch](https://microsoft.github.io/monaco-editor/monarch.html) looks like (without trying it) a conceptual hybrid between Tree-sitter and Sublime.

After a week of struggling with microsoft/vscode-textmate#32, an old version of Onigurama, and practically nonexistent documentation apart from [a blog post](https://www.apeth.com/nonblog/stories/textmatebundle.html) from 2014 ... could we pretty please with a cherry on top have an update on this issue?

**Comment by texastoland** [comment]
Continued from https://github.com/microsoft/vscode-textmate/issues/117#issuecomment-920925771:

> [@jeff-hykin] I spent years on a library (which I finally published just last week) to make it way less painful.

If it works for you that's great. For me most of your use case is solved by using YAML instead of JSON (like Sublime but it's frustrating that there's a [compile step](https://github.com/pedro-w/vscode-tmlanguage/releases) for Code) and the metaprogramming facilities of [embedding match content in scope names](https://macromates.com/blog/2011/format-strings/#the-case-for-replacements) (thenusing them like CSS classes to inject other grammars) or YAML 1.1 [merge keys](http://blogs.perl.org/users/tinita/2019/05/reusing-data-with-yaml-anchors-aliases-and-merge-keys.html).

How YAML works (syntax highlighting available for regexes):

```yaml
scopeName: inline.template-fsharp-highlight.reinjection
injectionSelector: "L:meta.embedded"
patterns:
  - name: string.quoted.triple.template.substitution.fsharp
    contentName: meta.template.expression.fsharp
    begin: |
      (?x)    # Ignore whitespace
      (?<!\{) # Not after brace
      \{      # Literal brace
      (?!\{)  # Not before brace
    end: |
      (?x)
      (?<!})
      }
      (?!})
    captures:
      0:
        name: keyword.symbol.fsharp
    patterns:
      - include: source.fsharp
```

I've seen at least 2 projects that rolled their own grammar generators (the original ReScript syntax and your own Better Shell Syntax). There's even a more interesting compiler (currently [documented](https://eeyo.io/iro/documentation/), [online REPL](https://eeyo.io/iro/), and [CLI](https://github.com/c272/iro4cli) but [no extension yet](url)) to transpile an entirely new syntax with a Sublime-like stacking context into TextMate.

But to me it's all infuriating. Code is so progressive in so many ways but not only regressive in a core component of literally any text editor but now unresponsive about it. The https://github.com/microsoft/vscode-textmate project is on 1 hand daunting and on the other janky and impossible to discern whether it's due to TextMate's unspecified behavior or actually a bug.

Semantic tokens were a foundational step but not a solution. Most of the implementations connect them to their LSP. That's less performant than using Tree-sitter (#50140), leaves the burden on extension authors to provide a TextMate grammar when the LSP isn't available (like for a file outside a .NET project), and creates an inconsistent experience for end users in terms of coloring (whitespace significant languages like F# are most drastically affected) as well as performance.

In conclusion being silent about this hurts:

1. performance of arguably 1 of the primary functionalities of Code
2. extension authors who waste weeks reinventing the wheel because of an undocumented, outdated, insufficient, and slightly buggy tool
3. end user experience

**Comment by jasonwilliams** [comment]
@aeschli  do you know if there’s any current exploration into something like treesitter as a replacement for the textmate grammars we have today? It can’t just be left as it is indefinitely as it’s noticeably bad.

> But to me it's all infuriating. Code is progressive in so many ways but not only regressive in a core component of literally any text editor but now unresponsive about it.

> performance of arguably 1 of the primary functionalities of Code.

This is very true and it’s actually quite sad to see too.

It’s great VSCode has all of these fancy bells and whistles and more features than you can possibly need, but it seems to get the basics wrong when it comes to rendering the source code. On typescript projects I see the syntax highlighting kick in a few second after the code shows up, this is such a fundamental issue. I understand there’s also a desire to fully rely on LSP for code colouring, but this just adds extra latency like @texastoland mentioned above; you would definitely need some level of caching or stale-while-revalidate before falling back to LSP otherwise it’s no better than what we have today.

I understand anything around tokenisation requires a refactor and that’s most likely why no one wants to go near it but how long can that last really? Until [competition](https://zed.dev/) begins to narrow?

> Any additional tokenizer is waiting on @alexdima's https://github.com/microsoft/vscode/issues/77140.

his last response to that thread was almost 3 years ago so I think it’s a dead end. It’s yet another thread where the maintainers have gone silent on the issue.

I did include tree-sitter in my post around VSCode performance as a whole https://jason-williams.co.uk/speeding-up-vscode-extensions-in-2022

**Comment by jasonwilliams** [comment]
I've had a go at integrating a different service (alongside the textmate one) which supports tree-sitter. So far it loads up fine but there's some issues having it properly instantiate tree-sitter. I think this is to do with the security policies in place. 

I think its possible to have a Tree Sitter Service which can emit tokens (similar to the textmate service) and have higher-up services use that instead. Or have them use the tree sitter API wrapped in a service (for queries etc)

If anyone is interesting in helping there's a PR here:
https://github.com/microsoft/vscode/pull/147648

**Comment by texastoland** [comment]
> Until [competition](https://zed.dev/) begins to narrow?

Switching to Zed 🤷🏼‍♂️

**Comment by reteps** [comment]
(tangentially related) It would be great if we can use these grammars within language servers as well (https://github.com/microsoft/vscode-discussions/discussions/2656)
